### PR TITLE
Java: Fast DafnySequence <-> array conversion

### PR DESCRIPTION
**Pull request by lukemaurer**
Also did some refactoring to make sure that instanceof checks never
accidentally see a ConcatDafnySequence rather than the underlying class.

(Note that there's also an addition of support for unsigned byte arrays, which unfortunately is doomed to be undone by the removal of UByte (see #522). Which is to say, most of the changes from the beginning of (the new) AbstractByteArrayDafnySequence to the beginning of StringDafnySequence are about to be reversed.)

### CODE REVIEW

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 8334c995707988bf0cd0f8bde22b4e87b2fcfc27
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=490ca802-3df7-4f2b-87d3-06bf1643730f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: b22e387e3c6646d7ee4ef9240a25275bf0e39da0
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=86847e5c-f096-41c1-9ac2-853a8169c92e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java [hunk 1]
```diff
@@ -55,6 +89,39 @@
         return new ArrayDafnySequence<>(values);
     }
 
+    @SuppressWarnings("unchecked")
+    public T[] toArray(Class<T> type) {
```
Comment (robin-aws) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: This could be a lot more efficient with the help of a `copyIntoArray(T[] array, int index)` helper method which could be overridden by the various subtypes to avoid the intermediate List<T> where possible (assuming that still makes sense after your subsequent PRs)

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java [hunk 2]
```diff
@@ -265,11 +332,24 @@ public boolean contains(T t) {
...
+        DafnySequence<T> other = (DafnySequence<T>) obj;
+        return this.equalsNonLazy(other.force());
+    }
+
+    /**
+     * Compare for equality to the given sequence, assuming that it is not
+     * null, not == to this, and not lazy.
+     */
+    protected boolean equalsNonLazy(NonLazyDafnySequence<T> other) {
+        return asList().equals(other.asList());
```
Comment (robin-aws) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: Same comment here as in `toArray()`

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java [hunk 3]
```diff
@@ -28,6 +29,14 @@
         return new ArrayDafnySequence<>(elements.clone());
     }
 
+    /**
+     * Return a sequence backed by the given array without making a defensive
+     * copy.  Only safe if the array never changes afterward.
+     */
+    public static <T> DafnySequence<T> unsafeWrapArray(T[] elements) {
```
Comment (robin-aws) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: I totally appreciate the need for doing this, but I wonder if we can at least have an `unsafe` boolean on `ArrayDafnySequence` to help debug things if they go sideways.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java [hunk 4]
```diff
@@ -295,9 +375,25 @@ public String verbatimString(){
     public HashSet<T> UniqueElements() {
```
Comment (robin-aws) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: Drive-by nit: probably should be `uniqueElements()`?

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java (hunk 1 above)
Comment (lukemaurer) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: Since `asList()` is just a view object, there isn't really an intermediate list (though of course bulk copying methods are always fast).  But yeah, this is something that changes again with the unboxed primitives PR (and there the subclasses do use bulk operations; CTTOI `toArray()` could probably be abstract in the superclass at that point).

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java (hunk 3 above)
Comment (lukemaurer) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: Like, a flag that that just records whether the array was constructed using `unsafeWrapArray`, for debugging purposes?  Sure, that sounds reasonable.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java (hunk 4 above)
Comment (lukemaurer) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: It's uppercase because the name is hardcoded in the `Compiler` base class (`dafny.Helpers.Quantifier()` is the same way), which of course follows C# naming sensibilities.  (I suppose we could also offer a lowercase version for native code.)

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: b7130ff8811bdf5fc7c598ac533be45b2bacfafa
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a4f091ba-41e9-455f-9e65-009f3689968d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: df34d0e5451fbdfc5ace82f729c371596bc674c6
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=ea84ecd8-5a4d-4634-b854-65658e33f7b9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: ebdd77d01b2d211fc95e0206e0caf76a42f87e8b
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=2ee4be68-d7ff-4c73-a57c-12dfa5df886c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java (hunk 2 above)
Comment (lukemaurer) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: This one is actually necessary as a fallback.  All the overrides expect an argument of a certain type, and if the argument isn't that type, we can't do any better than iterating over both sequences, which is what this line really does.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java (hunk 4 above)
Comment (robin-aws) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: Okay that's definitely a deeper issue than I'd want to fix now then. :)

**Comment by RustanLeino** [comment]
I says "All checks have failed", but the Build Logs indicate that the test suite passed. I'm going to merge this PR, especially since the subsequence PR (which I will review next) is passing the test suite.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java [hunk 5]
```diff
@@ -396,10 +396,17 @@ public String verbatimString(){
 final class ArrayDafnySequence<T> extends NonLazyDafnySequence<T> {
     // not T[] because generics and arrays don't mix
     private Object[] seq;
+    @SuppressWarnings("unused")
+    private boolean unsafe; // for debugging purposes
```
Comment (robin-aws) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java]: I'd actually suggest an `isUnsafe()` method on the interface (or perhaps something more precise) which you can call in a unit test.
